### ISSUE TITLE
11.1-Adversarial Attack

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#65]
### 人類的惡意
要把這些 Network 用在真正應用上光是它們正確率高是不夠的
他們需要能夠應付來自人類的惡意

有時候你的 Network是為了要偵測一些有惡意的行為
它要偵測的對象(人類)則會去想辦法騙過 Network

Network要在有人試圖想要欺騙它的情況下也得到高的正確率
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/336c9900-72b3-4939-bd89-7b4bcf71d00d)

舉例來說 你會用 Network 來做偵測一封郵件它是不是垃圾郵件
對於一個垃圾郵件的發信者而言
他也會想盡辦法避免他的郵件被分類為垃圾郵件

我們先來看一個真正的例子:一個影像辨識的系統
給它一張照片, 它可以告訴我們說這張照片屬於什麼樣的類別

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/d562060a-2644-4090-9ef4-84fefad253c4)

在這張照片上面加入一個非常小的雜訊(最好小到人肉眼沒有辦法看出來)
一張照片可以被看作是一個非常長的向量
我們在這個非常長的向量每一個維度都加入一個小小的雜訊
有被加雜訊的照片叫做 Attacked Image
還沒有被加雜訊的照片(被攻擊的照片)一般就叫做 Benign Image

攻擊大致上可以分成兩種類型

1.沒有目標的攻擊
原來的答案是貓->讓 Network 的輸出不是貓

2.有目標的攻擊
原來的答案是貓->讓 Network 的輸出是某個目標(例如海星)

實際例子
這邊用 50 層的 ResNet Network (並不是一個很弱的 Network)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/7ff61f4d-24a6-4786-ae8f-e63dfaaffa75)

Benign Image -> Tiger Cat (信心的分數是 0.64)
信心分數就是做完那個 Soft Mask 以後得到的那個分數
Attacked image -> Star Fish (信心的分數是 1.00)

把這兩張照片做一下相減並放大 50 倍
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/53620103-8c15-4c4a-82fd-06a3ab817442)

這兩張照片人根本看不出差異, 但確實有些不一樣

人類的惡意攻擊是非常可怕的
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/ee1d138a-fd5e-4c4a-bc61-47a661eb3d28)

它可不是一個非常爛的 Network
如果你加入的只是一般的雜訊它並不一定會犯錯
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/307cfae1-e57c-4e05-a77d-3cad5f4a67c1)

___________________

# 攻擊是如何做到的

Network, = Function, f
Function 輸入是一張圖片 =  x0
Function 輸出是一個 Distribution
分類的結果 = y0
正確的答案 =  ŷ (ŷ 其實是一個 One-Hot Vector)

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/98c588e4-6eaa-4a75-8e56-90158b5262a8)

Network 的參數就是固定的(上線的都是訓練好的module)

### 如何找出 Non-Targeted Attack 的雜訊

找一張新的圖片(用 x 來表示)
當我們把 x 丟到這個 Network f 的時候它的輸出是 y
我們希望 y 跟 ŷ 它的差距越大越好
我們一樣要解一個 Optimization 的問題
這跟我們訓練 Network其實是非常類似的
定一個 Loss Function = y 跟 ŷ 之間的差距取一個負號

一般在做這個 Classification 的時候
我們訓練的目標 y 跟 ŷ都是看它的 Cross Entropy

e(y, ŷ) = y 跟 ŷ 之間的 Cross Entropy
但是我們希望這個 Cross Entropy 越大越好
所以我們今天在 Cross Entropy 前面加一個負號
這個負的 Cross Entropy 就是我們的 Loss (我們期望這個 Loss 越小越好)
我們希望找到一個 x 可以讓 L(x) 越小越好
L(x) 越小 就代表說 y 跟 ŷ它們的 Cross Entropy 越大
也就是 y 跟 ŷ 它們的距離越大

### 如何找出 Targeted Attack 的雜訊

先設定好我們的目標 =  y target (y target 也是一個 One-Hot Vector)
假設 y target 是一個 Fish

希望 y 跟 ŷ 越遠越好+ y 跟 y target 越近越好
y 跟 ŷ 越遠越好 = 負的 y 跟 ŷ 之間的 Cross Entropy
y 跟 y target 越近越好 = y 跟 y targe 之間的 Cross Entropy

你希望找一個 x 去 Minimize 這個 Loss

我們其實還期待說我們加入的雜訊越小越好(也就是我們新找到的圖片可以欺騙過 Network 的圖片)
跟原來的圖片要越相近越好 = x 跟 x0 要越近越好
所以我們在解這個Optimization 的 Problem 的時候
我們還會多加入一個限制 : d(x0,x)≦ε

我們希望 x 跟 x0 之間的差距小於某一個 Threshold ε
ε通常就是根據人類的感知能力來決定
如果 x0 跟 x 之間的差距大於  ε ->人類就會看到這個雜訊

### d(x0,x)≦ε

假設 x和x0 是一個向量(圖片)
Δx = x-x0
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/3f47c2ab-5277-4f95-9e92-21efa5ebacb8)

L2-Norm 的定義你要開根號也是可以的

L-Infinity哪一個維度它的絕對值最大->那這一個值就是L-Infinity
看起來應該是 L-Infinity 比較接近人類的感知能力->所以在作業裡面是用 L-Infinity

如果我們今天要攻擊的對象其實是一個跟語音相關的系統
那就不見得是 L2 跟 L-Infinity 了-> Domain Knowledge

### 解 Optimization 

先把限制 d(x0,x)≦ε 拿掉->跟 Train 一個模型其實沒有什麼差別
就跟一般的 Gradient Descent 是一模一樣的
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/24684cde-7b54-4be7-8b5c-20f7ce86f3a1)

接下來我們得把 Constraint 加進去->同時考慮 x0 跟 x 之間的差距
如果你 Update 完你的參數以後發現 x0 跟 xt差距大於 ε 以後->把 xt 改回符合限制
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/cac1362f-e163-440c-b0b0-55a3fd66af30)

最簡單的 Attack 的方法 - ### FGSM

Fast Gradient Sign Method = FGSM
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/ca126b4b-e473-4ba0-8b6e-aba4cc2ef095)

FGSM它就像是一個一拳超人一樣只用一擊
FGSM只 Update 一次參數

加了 Sign 以後 g 這個 Vector 裡面要嘛是 1 要嘛是 -1
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/d198b0d1-53b3-4f4b-9367-db98e86ec68c)

Learning Rate 就設 ε

你攻擊完以後你一定落在這個藍色框框的四個角落的地方

多跑幾個 Iteration 結果會更好->你要跑幾個 Iteration高興都是你自己設
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/8e6b692d-2afa-4c84-a17a-408fe74ccc57)

多跑幾個 Iteration 的壞處是你有可能一不小心就出界->把它拉回來
